Logging proto3

josguil opened the pull request:
Proposed change:

Logging class will be replaced by particular logging classes depending
on the component doing the logging.  Instead of doing
"Logging.Enter(Logging.Socket, socket, "method", "parameters")" the new
call will look like this: "SocketsLogging.Enter(socket,  "method",
"parameters")".

This is done so each project don't have to include every logging related
file and only the ones related to that component. A few logging lines
are replaced using this proposed new system in this prototype.

The original logging class is being left for now for 3 reasons:

-So old calls using shims can still work (at least until they get
replaced by the new format, which will be the next step after this
prototype is approved).

-Some common methods related to hashing are in logging.cs, until a
better place is found for them.

-The enum value related to the eventIDs will be there also until a
better fit place is found.

josguil committed:
Logging Prototype version 3

Proposed change:

Logging class will be replaced by particular logging classes depending
on the component doing the logging.  Instead of doing
"Logging.Enter(Logging.Socket, socket, "method", "parameters")" the new
call will look like this: "SocketsLogging(socket,  "method",
"parameters")".

This is done so each project don't have to include every logging related
file and only the ones related to that component. A few logging lines
are replaced using this proposed new system in this prototype.

The original logging class is being left for now for 3 reasons:

-So old calls using shims can still work (at least until they get
replaced by the new format, which will be the next step after this
prototype is approved).

-Some common methods related to hashing are in logging.cs, until a
better place is found for them.

-The enum value related to the eventIDs will be there also until a
better fit place is found.

josguil committed:
Logging Prototype 3

Proposed change:

Logging class will be replaced by particular logging classes depending
on the component doing the logging.  Instead of doing
"Logging.Enter(Logging.Socket, socket, "method", "parameters")" the new
call will look like this: "SocketsLogging(socket,  "method",
"parameters")".

This is done so each project don't have to include every logging related
file and only the ones related to that component. A few logging lines
are replaced using this proposed new system in this prototype.

The original logging class is being left for now for 3 reasons:

-So old calls using shims can still work (at least until they get
replaced by the new format, which will be the next step after this
prototype is approved).

-Some common methods related to hashing are in logging.cs, until a
better place is found for them.

-The enum value related to the eventIDs will be there also until a
better fit place is found.

josguil commented:
@cipop , @davidsh PTAL.

CIPop commented on src/System.Net.Sockets/src/Resources/Strings.resx:
```diff
@@ -273,4 +273,7 @@
   <data name="ArgumentOutOfRange_Bounds_Lower_Upper" xml:space="preserve">
     <value>Argument must be between {0} and {1}.</value>
   </data>
+  <data name="event_SocketAccepted" xml:space="preserve">
```
Where is this used?

CIPop commented on src/System.Net.Requests/src/System.Net.Requests.csproj:
```diff
@@ -42,8 +42,17 @@
-    <Compile Include="$(CommonPath)\System\Net\Logging.cs">
-      <Link>Common\System\Net\Logging.cs</Link>
+    <Compile Include="$(CommonPath)\System\Net\Shims\TraceSource.cs">
+      <Link>Common\System\Net\Shims\TraceSource.cs</Link>
```
I presume this is needed to support the old Logging.cs dependencies and it will be deleted in the final version?

josguil replied:
Yes, exactly.

CIPop commented on src/Common/src/System/Net/Logging/SocketsLogging.cs:
```diff
@@ -0,0 +1,294 @@
+
+namespace System.Net
+{
+    internal static class SocketsLogging
```
We can actually share the common code. The [Working Effectively with Legacy Code](http://www.amazon.com/Working-Effectively-Legacy-Michael-Feathers/dp/0131177052) book describes the technique as code seaming.
Make Logging partial and refactor the Logging.cs code such that it's written as a template class. The actual definition of the eventSource will need to remain particular to each component (e.g. Logging.Sockets.cs).

CIPop commented on src/Common/src/System/Net/Logging/SocketsLogging.cs:
```diff
@@ -0,0 +1,294 @@
+            s_socketsEventSource.DebugMessage(GetManagedThread(), msg);
+        }
+
+        internal static void PrintInfo(object obj, string msg)
```
I presume these will be removed right?

CIPop commented on src/Common/src/System/Net/Logging/WebEventSource.cs:
```diff
@@ -0,0 +1,108 @@
+                EventData* dataDesc = stackalloc EventSource.EventData[4];
+
+                dataDesc[0].DataPointer = (IntPtr)arg1Ptr;
+                dataDesc[0].Size = (managedThreadID.Length + 1) * 2; // Size in bytes, including a null terminator. 
```
What's the *2 constant?

josguil replied:
The size of each character in a string.

josguil committed:
Adding a template common class and more examples.

LoggingUtilities serves as the proposed classed where all code common
for logging will be. It is a generic class so it can call some of the
logging methods common to all components.

Added some examples of how printinfo can be substituted for new methods.

josguil commented:
I added a LoggingUtilities generic (template) class that will have all the common code between logging events.
In order to do that, each EventSource has to implement an interface (ILogging, defined for now in the LoggingUtilities class) with the common logging methods (as  FunctionStart and FunctionStop). 
Since interfaces can not have any static elements, and our eventsources follow the singleton pattern,  it was not possible to have any code related to initialization in the LoggingUtilities class. This includes validation, checking for initialization, and the property On. Those will need to remain repeated in each Logging class.

@CIPop  @davidsh  PTAL

josguil committed:
Deleting Facade classes ("SocketsLoggin.cs")

As discussed, facades classes are not necessary anymore, since now there
are no common event methods between eventsources. Enter, Exit and
CriticalEvents will be exclusive to NetEventSource. This will be the
only event methods present in NetEventSource (previously
WebEventSource). Older event methods that were in WebEventSource will be
moved to a new SecurityEventSource since Security is the only consumer
of those events.

In this change:

-Adding the proposed Enter method in NetEventSource.

-Adding an example of Accepted event method in the SocketsEventSource
class.

-Adding HashLogging class with some common hashing methods.

-Removing facades.

-Changing enums used as event ids for const int types.

josguil commented:
Only look at the changes done under Net/Logging and System.Net.Sockets. I didn't have time to clean the code under other projects.

CIPop commented on src/Common/src/System/Net/Logging/NetEventSource.cs:
```diff
@@ -0,0 +1,106 @@
+        private const int FUNCTIONSTART_ID = 1;
+
+        private static NetEventSource s_log = new NetEventSource();
+        private NetEventSource() { }
```
Nit: missing new lines.

CIPop commented on src/Common/src/System/Net/Logging/NetEventSource.cs:
```diff
@@ -0,0 +1,106 @@
+        [NonEvent]
+        internal static void Enter(object obj, string method, object paramObject, Component componentType)
+        {
+            if (!s_log.IsEnabled())
```
Nit: formatting (throughout the function).

CIPop commented on src/Common/src/System/Net/Logging/NetEventSource.cs:
```diff
@@ -0,0 +1,106 @@
+            int callerHash;
+            string parametersName = "";
+            int parametersHash = 0;
+            if (obj is string)
```
Why would this be a string?

josguil replied:
Previously we had 4 enter methods, some of them were receiving a string.

CIPop commented on src/Common/src/System/Net/Logging/NetEventSource.cs:
```diff
@@ -0,0 +1,106 @@
+                EventData* dataDesc = stackalloc EventSource.EventData[6];
+
+                dataDesc[0].DataPointer = (IntPtr)arg1Ptr;
+                dataDesc[0].Size = (callerName.Length + 1) * sizeof(short); // Size in bytes, including a null terminator. 
```
Why sizeof(short)? `arg1Ptr` is a `char`.

josguil replied:
Done

CIPop commented on src/Common/src/System/Net/Logging/NetEventSource.cs:
```diff
@@ -0,0 +1,106 @@
+            
+            fixed (char* arg1Ptr = callerName, arg2Ptr = method, arg3Ptr = parametersName)
+            {
+                EventData* dataDesc = stackalloc EventSource.EventData[6];
```
You should define a name for `6` and reuse it below.

CIPop commented on src/Common/src/System/Net/Logging/NetEventSource.cs:
```diff
@@ -0,0 +1,106 @@
+                dataDesc[5].DataPointer = (IntPtr)(&componentType);
+                dataDesc[5].Size = sizeof(int);
+
+                WriteEventCore(FUNCTIONSTART_ID, 6, dataDesc);
```
Define a constant for `6` and reuse it here (see above).

CIPop commented on src/Common/src/System/Net/Logging/NetEventSource.cs:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        [NonEvent]
+        internal static void Enter(object obj, string method, object paramObject, Component componentType)
```
For clarity, I believe that Component should be the first parameter.

CIPop commented on src/Common/src/System/Net/Logging/NetEventSource.cs:
```diff
@@ -0,0 +1,106 @@
+            public const EventKeywords FunctionEntryExit = (EventKeywords)0x0004;
+        }
+
+        public enum Component
```
Nit: `ComponentType` instead of `Component`

CIPop commented on src/Common/src/System/Net/Logging/SocketsEventSource.cs:
```diff
@@ -0,0 +1,62 @@
+
+    [EventSource(Name = "Microsoft-System-Net-Sockets",
+        Guid = "e03c0352-f9c9-56ff-0ea7-b94ba8cabc6b",
+        LocalizationResources = "FxResources.System.Net.Sockets.SR")]
```
Is this needed to make this work? I'm afraid that it will silently stop working if someone decides to rename the resource file.

CIPop commented on src/Common/src/System/Net/Logging/SocketsEventSource.cs:
```diff
@@ -0,0 +1,62 @@
+
+                EventData* dataDesc = stackalloc EventSource.EventData[3];
+                dataDesc[0].DataPointer = (IntPtr)arg1Ptr;
+                dataDesc[0].Size = (localEp.Length + 1) * sizeof(short); // Size in bytes, including a null terminator. 
```
Same observation as above: sizeof(char) probably?

CIPop commented on src/System.Net.Http/src/Resources/Strings.resx:
```diff
@@ -354,4 +354,7 @@
     <value>The server returned an invalid or unrecognized response.</value>
   </data>
-</root>
+  <data name="net_log_exception" xml:space="preserve">
```
Is this used?

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -86,10 +86,10 @@ public Socket(SocketType socketType, ProtocolType protocolType)
             if (s_loggingEnabled)
             {
-                Logging.Enter(Logging.Sockets, this, "Socket", addressFamily);
+                NetEventSource.Enter(this, "Socket", addressFamily, NetEventSource.Component.Socket);
```
Would be easier if you moved the Component type first as the transformation will be 1:1:  Logging.Sockets -> NetEventSource.Component.Socket; Logging->NetEventSource and the rest remains the same.

josguil committed:
Adding skeleton for proposed events

HttpEventSource and, SecurityEventSource skeletons are added, with a few
events methods as example.

Also, addressing feedback in Git:

-ComponentType now is the first parameter.
-Changing sizeof(short) to sizeof (char)
-Deleting some not used elements from previous iterations
-Renaming enum Component to ComponentType
-Defining a constant for the size of the eventData array.

josguil committed:
Adding a few changes missed in previous iteration

Changing some shorts to chars and using constant for the size of the
array,

josguil committed:
Cleaning Code

Reverting back unused changes

josguil committed:
Undoing some changes not longer used.

A few changes were missing in last commit.

josguil commented:
@CIPop  I cleaned the code, and put the skeleton and examples we talked about.
